lib/CL/pocl_llvm_api.cc

```diff
@@ -534,54 +534,112 @@ static int pocl_llvm_get_kernel_arg_metadata(const char* kernel_name,
                                              llvm::Module *input,
                                              cl_kernel kernel)
 {
-
-  // find the right kernel in "opencl.kernels" metadata
-  llvm::NamedMDNode *opencl_kernels = input->getNamedMetadata("opencl.kernels");
   llvm::MDNode *kernel_metadata = NULL;
+  int bitcode_is_spir = input->getTargetTriple().find("spir") == 0;
+  SmallVector<std::pair<std::string, llvm::MDNode*>, 6> llvm39_kernel_meta;
+  bool llvm_older_than_3_9 = false;
+#if defined(LLVM_OLDER_THAN_3_9)
+  llvm_older_than_3_9 = true;
+#endif
 
-  // Not sure what to do in this case
-  if (!opencl_kernels || opencl_kernels->getNumOperands() == 0) return -1;
-
-  for (unsigned i = 0, e = opencl_kernels->getNumOperands(); i != e; ++i) {
-    llvm::MDNode *kernel_iter = opencl_kernels->getOperand(i);
-
-    llvm::Function *kernel_prototype = 
-      llvm::cast<llvm::Function>(
-        dyn_cast<llvm::ValueAsMetadata>(kernel_iter->getOperand(0))->getValue());
-    std::string name = kernel_prototype->getName().str();
-    if (name == kernel_name) {
-      kernel_metadata = kernel_iter;
-      break;
+  llvm::NamedMDNode *opencl_kernels = input->getNamedMetadata("opencl.kernels");
+  if (opencl_kernels) {
+    // find the right kernel in "opencl.kernels" metadata
+
+    // Not sure what to do in this case
+    if (!opencl_kernels || opencl_kernels->getNumOperands() == 0) return -1;
+
+    for (unsigned i = 0, e = opencl_kernels->getNumOperands(); i != e; ++i) {
+      llvm::MDNode *kernel_iter = opencl_kernels->getOperand(i);
+
+      llvm::Function *kernel_prototype =
+        llvm::cast<llvm::Function>(
+                                   dyn_cast<llvm::ValueAsMetadata>(kernel_iter->getOperand(0))->getValue());
+      std::string name = kernel_prototype->getName().str();
+      if (name == kernel_name) {
+        kernel_metadata = kernel_iter;
+        break;
+      }
     }
+    assert(kernel_metadata && "kernel NOT found in opencl.kernels metadata");
+  }
+  // LLVM 3.9 does not use opencl.kernels meta, but kernel_arg_* function meta
+  // Store meta and names in a vector so we can use rest of the code as is
+  else {
+    for (llvm::Module::iterator i = input->begin(), e = input->end();
+         i != e; ++i) {
+      if (i->getMetadata("kernel_arg_access_qual")
+          && i->getName() == kernel_name)
+        {
+          //printf("found kernel with name\n");
+          //i->getMetadata("kernel_arg_type")->dump();
+
+          llvm39_kernel_meta.push_back
```

Comment:
Why a new runtime variable? Why not use the LLVM_OLDER_THAN_3_9 macro with different code paths?